### PR TITLE
Add truncation support to SpriteText

### PR DESCRIPTION
**Pull request by ThePyrotechnic**
Unsure whether both Truncate and TruncateWithEllipsis are desired

- Prerequisite for ppy/osu#2819

---

Add Truncate and TruncateWithEllipsis to SpriteText

### CODE REVIEW

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 1]
```diff
@@ -434,12 +472,38 @@ private void computeCharacters()
                     Vector2 scaledTextureSize = textureSize * TextSize;
 
                     // Check if we need to go onto the next line
-                    if (AllowMultiline && currentPos.X + glyphSize.X >= maxWidth)
+                    if (AllowMultiline)
                     {
-                        currentPos.X = Padding.Left;
-                        currentPos.Y += currentRowHeight + spacing.Y;
-                        currentRowHeight = 0;
+                        if (currentPos.X + glyphSize.X >= maxWidth)
```
Comment (ProgrammaticNajel) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Why place this expression in another if statement instead of combining it with the if statement above?

Comment (ThePyrotechnic) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Consider this case:
```c#
AllowMultiline == true
currentPos.X + glyphSize.X >= maxWidth == false

TruncateWithEllipsis == true
currentPos.X + glyphSize.X + ellipsisWidth >= maxWidth == true
```
The user has indicated that they wish to `AllowMultiline`, but currently the text itself hasn't overflowed. _But_ the text with ellipsis will overflow, meaning that (if the statements are combined)  the text will get truncated even though `AllowMultiline` is `true`. In my setup (not combined) I have `AllowMultiline` overriding both `Truncate` and `TruncateWithEllipsis`. It is not possible to have both truncation and mulitline wrapping simultaneously, so one must be dominant.

In other words, this structure says in English: "Regardless of whether a line needs to be wrapped, do not try to truncate if `AllowMultiline` is true".

Note that the same scenario cannot occur between `TruncateWithEllipsis` and `Truncate` since it is not possible for:
```c#
(currentPos.X + glyphSize.X + ellipsisWidth) < (currentPos.X + glyphSize.X)
```

Comment (ProgrammaticNajel) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Now that I think about it, the other case is in an `else if`, so it does make sense. Thank you for answering.

**Comment by peppy** [comment]
I've refactored your implementation to 

- Remove the second `bool` (`TruncateWithEllipsis`) in favour of a customisable ellipsis string.
- Avoid code duplication by calling the same function to add characters
- Fix a potential preceding space appearing before the ellipsis.
- Use the ellipsis character by default instead of '...'

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 2]
```diff
@@ -233,6 +234,36 @@ public bool UseFullGlyphHeight
...
+        private string ellipsisString = "…";
+
+        protected string EllipsisString
+        {
+            get => ellipsisString;
+            set
+            {
+                if (ellipsisString == value) return;
+                ellipsisString = value;
+                invalidate();
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: `invalidate(true)`, I guess changing the ellipsis string wasn't tested

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 3]
```diff
@@ -233,6 +234,36 @@ public bool UseFullGlyphHeight
...
+            {
+                if (truncate == value) return;
+                truncate = value;
+                invalidate(true);
+            }
+        }
+
+        private string ellipsisString = "…";
+
+        protected string EllipsisString
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: I don't think this really warrants `protected`.

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 4]
```diff
@@ -386,68 +417,77 @@ private void computeCharacters()
                     return;
 
                 float maxWidth = float.PositiveInfinity;
+                float ellipsisWidth = 0;
+
                 if (!requiresAutoSizedWidth)
+                {
                     maxWidth = ApplyRelativeAxes(RelativeSizeAxes, new Vector2(base.Width, base.Height), FillMode).X - Padding.Right;
+                    ellipsisWidth = Truncate ? EllipsisString.Sum(c => getCharacterSize(c, true, out _).X) : 0;
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: What about `Spacing`?

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 5]
```diff
@@ -386,68 +417,77 @@ private void computeCharacters()
...
+                            currentPos.X = Padding.Left;
+                            currentPos.Y += currentRowHeight + spacing.Y;
+                            currentRowHeight = 0;
                         }
-
-                        textureSize = new Vector2(size);
                     }
-                    else
+                    // Check if this character would take us past the max width if we were to add ellipsis
+                    else if (truncate && currentPos.X + glyphSize.X + ellipsisWidth >= maxWidth)
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Doesn't this mean that it's potentially going to show ellipsis even when there's no more text available? I.e displaying the entire string would not exceed the width in the first place.

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 6]
```diff
@@ -386,68 +417,77 @@ private void computeCharacters()
...
+
                 foreach (var character in displayedText)
+                    if (!addCharacter(character, Truncate))
+                        break;
+
+                bool addCharacter(char character, bool truncate)
                 {
-                    bool useFixedWidth = Font.FixedWidth && UseFixedWidthForCharacter(character);
+                    // don't apply fixed width as we need the raw size to compare with glyphSize below.
+                    Vector2 size = getCharacterSize(character, false, out Texture texture);
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Really would like to see this called texturesize again, size is really uninformative.

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 3 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: you thinking public?

Comment (smoogipoo) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Yes

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 5 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: good catch

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 6 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: i renamed it because it's no longer the raw texture size (it is scaled to the `FontSize`). sure you want it to be `textureSize`?

Comment (smoogipoo) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: It was called exactly scaledTextureSize before for that reason.

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 5 above)
Comment (smoogipoo) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Also this condition doesn't take Spacing into account.

Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: After refactor spacing should not be included in this condition (just as a sidenote).

**Comment by peppy** [comment]
Blocking pending further refactoring.

**Comment by peppy** [comment]
For future reference:

image

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 7]
```diff
@@ -175,6 +176,9 @@ public bool AllowMultiline
                 if (allowMultiline == value)
                     return;
 
+                if (value && Truncate)
+                    throw new InvalidOperationException($"Cannot use {nameof(Truncate)} and {nameof(AllowMultiline)} at the same time.");
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Why is this one throwing an exception while `Truncate` sets the other to false?

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 8]
```diff
@@ -239,6 +243,44 @@ public bool UseFullGlyphHeight
             }
         }
 
+        private bool truncate;
+
+        /// <summary>
+        /// True if the text should be truncated when it exceeds a specified <see cref="Width"/>.
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]:
```suggestion
        /// True if the text should be truncated when it exceeds the <see cref="DrawWidth"/> of this <see cref="SpriteText"/>.
```

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 9]
```diff
@@ -239,6 +243,44 @@ public bool UseFullGlyphHeight
...
+                    AllowMultiline = false;
+
+                truncate = value;
+                invalidate(true);
+            }
+        }
+
+        private string ellipsisString = "…";
+
+        public string EllipsisString
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: XMLDoc

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 10]
```diff
@@ -239,6 +243,44 @@ public bool UseFullGlyphHeight
...
 
+        private bool truncate;
+
+        /// <summary>
+        /// True if the text should be truncated when it exceeds a specified <see cref="Width"/>.
+        /// <remarks>
+        /// Has no effect if no <see cref="Width"/> or custom sizing is set.
+        /// If enabled, <see cref="AllowMultiline"/> will be disabled.
+        /// </remarks>
+        /// </summary>
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Generally, [remarks](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/remarks) go outside of summaries, not nested.

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 11]
```diff
@@ -386,95 +431,171 @@ private void computeCharacters()
...
+                if (requiresAutoSizedWidth)
+                    base.Width = currentPos.X + Padding.Right;
+                if (requiresAutoSizedHeight)
+                    base.Height = currentPos.Y + Padding.Bottom;
+
+                isComputingCharacters = false;
+                charactersCache.Validate();
+            }
+
+            int getTruncationLength()
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Move these into their own methods. We should generally avoid using local functions in hot paths since they allocate a class:

image

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 7 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Because `AllowMultiLine` is on by default. felt correct to me

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 11 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Where are you seeing the "allocation of a class"? That screenshot shows that the class is doing allocations, but those are listed below as being results of the actual computations?

Comment (smoogipoo) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: In this case it seems that the linq method is transforming the local function into a class.

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 12]
```diff
@@ -386,95 +435,176 @@ private void computeCharacters()
...
-                float currentRowHeight = 0;
+                // Calculate period texture info outside the loop so that it isn't done per-character
+
+                if (truncate)
+                {
+                    Debug.Assert(!AllowMultiline);
+
+                    int displayCount = getTruncationLength();
+
+                    foreach (var character in displayedText.Take(displayCount))
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: image

This one also

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 7 above)
Comment (smoogipoo) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: I'd say it seems weird that a user can do:

```
Text.Truncate = true
```

with no worries but:

```
Text.Multiline = true
```

will crash...

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 13]
```diff
@@ -386,95 +434,176 @@ private void computeCharacters()
...
+
+            int getTruncationLength()
+            {
+                float trackingPos = Padding.Left;
+
+                float ellipsisLength = 0;
+                foreach (var c in EllipsisString)
+                    ellipsisLength += getCharacterSize(c, true, out _).X + spacing.X;
+
+                float availableWidth = maxWidth -= ellipsisLength;
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: What about `Padding.Right`?

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 14]
```diff
@@ -386,95 +434,176 @@ private void computeCharacters()
...
-                    Vector2 textureSize;
-                    Texture texture = null;
+                    // Scaled glyph size to be used for positioning.
+                    float glyphWidth = useFixedWidthForCharacter(character) ? constantWidth * Font.Size : scaledTextureSize.X;
 
-                    // Retrieve the texture + size
-                    if (char.IsWhiteSpace(character))
-                    {
-                        float size = useFixedWidth ? constantWidth : spaceWidth;
+                    if (trackingPos + glyphWidth >= availableWidth)
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: How about just `>`? `==` won't exceed the bounds.

File: osu.Framework/Graphics/Sprites/SpriteText.cs [hunk 15]
```diff
@@ -386,95 +434,176 @@ private void computeCharacters()
...
+                float availableWidth = maxWidth -= ellipsisLength;
+
+                int index = 0;
+                int lastNonSpaceIndex = 0;
 
                 foreach (var character in displayedText)
                 {
-                    bool useFixedWidth = Font.FixedWidth && UseFixedWidthForCharacter(character);
+                    // don't apply fixed width as we need the raw size to compare with glyphSize below.
+                    Vector2 scaledTextureSize = getCharacterSize(character, false, out Texture texture);
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Sprites/SpriteText.cs]: Is there a reason why this can't be `getCharacterSize(character, true, out Texture texture` instead, for this pass?

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 14 above)
Comment (smoogipoo) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Don't worry about this one - as discussed it won't be noticeable anyway.

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 15 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Well spotted

File: osu.Framework/Graphics/Sprites/SpriteText.cs (hunk 13 above)
Comment (peppy) [reply osu.Framework/Graphics/Sprites/SpriteText.cs]: Already considered at line 449
